gem/optimise.py

```diff
@@ -413,3 +480,555 @@ def aggressive_unroll(expression):
+        for of in optimal_factors:
+            if of in all_factors:
+                all_factors.remove(of)
+                factor_dict[of].append(all_factors)
+                break
+        else:
+            factor_lists.append(all_factors)
+    for of, factors in factor_dict.items():
+        # factors = list of lists representing sum of products
+        if len(factors) == 1:
+            # just one product
+            factor_lists.append([of] + factors[0])
+        else:
+            if len(arg_ind_flat) > 2:
+                # more argument indices to process
+                if len(set(of.free_indices) & set(arg_ind_flat)) != 1:
+                    raise AssertionError("this should not happen")
+                ind = (set(of.free_indices) & set(arg_ind_flat)).pop()
+                new_arg_ind_flat = tuple([i for i in arg_ind_flat if i != ind])
+                if len(arg_ind_flat) - len(new_arg_ind_flat) != 1:
+                    raise AssertionError("this should not happen")
+                factor_lists.append([of, cse(sumproduct_2_node(factors),
+                                             quad_ind, new_arg_ind_flat)])
+            else:
+                # remember to factorise factor list here
+                factor_lists.append([of, sumproduct_2_node(factorise(factors, quad_ind, arg_ind_flat))])
+    return sumproduct_2_node(factorise(factor_lists, quad_ind, arg_ind_flat))
+
+
+def gen_lex_sequence(items):
+    """
+    generate lexicographical order of permutations.
+    e.g. (i,j,k) -> [(i,), (i,j), (i,j,k), (j), (j,k), (k)]
+    :param items:
+    :return:
+    """
+    if not items:
+        return list()
+    result = gen_lex_sequence(items[1:])
+    return [(items[0],)] + [(items[0],) + x for x in result] + result
+
+
+@singledispatch
+def _reorder(node, self, indices):
+    """
+    Reorder Sum and Product to promote hoisting
+    :param node:
+    :param self:
+    :param indices:
+    :return:
+    """
+    raise AssertionError("cannot handle type %s" % type(node))
+
+
+_reorder.register(Node)(reuse_if_untouched_arg)
+
+
+@_reorder.register(IndexSum)
+def _reorder_indexsum(node, self, indices):
+    new_indices = tuple(indices)
+    for i in node.multiindex:
+        if i not in indices:
+            new_indices = (i,) + new_indices  # insert quadrature indice in front
+    new_children = [self(child, new_indices) for child in node.children]
+    return node.reconstruct(*new_children)
+
+
+@_reorder.register(Product)
+@_reorder.register(Sum)
+def _reorder_product_sum(node, self, indices):
+    _class = type(node)
+    all_factors = list(collect_terms(node, _class))
+    factor_lists = list()
+    for curr_indices in [()] + gen_lex_sequence(indices):
+        fl = list()
+        for f in all_factors:
+            if set(f.free_indices) == set(curr_indices):
+                fl.append(self(f, indices))
+        if fl:
+            factor_lists.append(fl)
```

Comment:
Also, it looks like if any node in the children has more free indices than the indices passed in, then you just drop that node on the floor!  Can you add an assertion that this is never the case?